packages/omezarr/src/zarr/cached-loading/store.ts

```diff
@@ -0,0 +1,256 @@
+    return `${keyStr} ${rangeStr}`;
+};
+
+type FetchStoreOptions = {
+    overrides?: RequestInit;
+    useSuffixRequest?: boolean;
+};
+
+class CacheableByteArray implements Cacheable {
+    #arr: Uint8Array;
+
+    constructor(arr: Uint8Array) {
+        this.#arr = arr;
+    }
+
+    destroy() {}
+
+    sizeInBytes(): number {
+        return this.#arr?.byteLength ?? 0;
+    }
+
+    buffer(): ArrayBufferLike {
+        return this.#arr.buffer;
+    }
+}
+
+type CacheKey = string;
+
+type TransferableRequestInit = Omit<RequestInit, 'body' | 'headers' | 'signal'> & {
+    body?: string;
+    headers?: Record<string, string>;
+};
+
+const copyHeaders = (headers: RequestInit['headers']): Record<string, string> | undefined => {
+    if (Array.isArray(headers)) {
+        const result: Record<string, string> = {};
+        headers.forEach(([key, val]) => {
+            // TODO is key, val the correct order here?
+            result[key] = val;
+        });
+        return result;
+    }
+    if (headers instanceof Headers) {
+        const result: Record<string, string> = {};
+        headers.forEach((val, key) => {
+            result[key] = val;
+        });
+        return result;
+    }
+    return headers;
+};
+
+const copyToTransferableRequestInit = (req: RequestInit | undefined): TransferableRequestInit => {
+    if (req === undefined) {
+        return {};
+    }
+    const updReq = { ...req };
+    delete updReq.signal;
+    delete updReq.window;
+    return { ...updReq, body: req.body?.toString(), headers: copyHeaders(req.headers) };
+};
+
+export type CachingMultithreadedFetchStoreOptions = {
+    maxBytes?: number | undefined;
+    numWorkers?: number | undefined;
+    maxFetches?: number | undefined;
+    fetchStoreOptions?: FetchStoreOptions | undefined;
+};
+
+type PromiseResolve<T> = (t: T) => void;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+// biome-ignore lint/suspicious/noExplicitAny: This is aligned with the standard Promise API
+type PromiseReject = (reason: any) => void;
+
+type PendingRequest<T> = {
+    resolve: PromiseResolve<T>;
+    reject: PromiseReject;
+    promise: Promise<T>;
+};
+
+export class CachingMultithreadedFetchStore extends zarr.FetchStore {
+    /**
+     * Maintains a pool of available worker threads.
+     */
+    #workerPool: WorkerPool;
+
+    /**
+     * Stores the current set of cached data that has been successfully
+     * fetched. This data is stored in raw byte array form so that it
+     * integrates properly with the Zarrita framework.
+     */
+    #dataCache: AsyncPriorityCache<CacheableByteArray>;
+
+    /**
+     * Maps cache keys to numeric times; the higher the time, the higher the priority.
+     *
+     * This effectively means that more frequently-requested items will be kept longer.
+     */
+    #priorityMap: Map<CacheKey, number>;
+
+    #pendingRequests: Map<CacheKey, PendingRequest<Uint8Array | undefined>>;
```

Comment:
this is the part felt fishy to me - AsyncPriorityCache has its own map of pending requests! that made me realize, we're doing most of its work here already, so much so that its not really helping